src/app/debugger.js

```diff
@@ -1,39 +1,132 @@
   this.debugger = new remix.ui.Debugger();
+  this.sourceMappingDecoder = new remix.util.SourceMappingDecoder()
   this.el.appendChild(this.debugger.render());
+  this.editor = editor;
+  this.switchToFile = switchToFile;
+  this.compiler = compiler;
+  this.cache = new Cache()
 
   var self = this;
   executionContextEvent.register('contextChanged', this, function (context) {
-    context = context === 'vm' ? 'VM' : context;
-    context = context === 'injected' ? 'EXTERNAL' : context;
-    context = context === 'web3' ? 'INTERNAL' : context;
     self.switchProvider(context);
   });
+
+  this.lastCompilationResult = null;
+  this.debugger.register('newTraceLoaded', this, function () {
+    self.cache.clear()
+    self.lastCompilationResult = self.compiler.lastCompilationResult;
+  });
+
+  this.debugger.register('traceUnloaded', this, function () {
+    self.lastCompilationResult = null;
+    self.removeCurrentMarker()
+    self.cache.clear()
+  });
+
+  this.editor.onChangeSetup(function () {
+    if (arguments.length > 0) { // if arguments.length === 0 this is a session change, we don't want to stop debugging in that case
+      self.debugger.unLoad()
+    }
+  });
+
+  // register selected code item, highlight the corresponding source location
+  this.debugger.codeManager.register('changed', this, function (code, address, index) {
+    this.debugger.sourceLocationTracker.getSourceLocation(address, index, self.lastCompilationResult.data.contracts, function (error, rawLocation) {
+      if (!error) {
+        if (!self.cache.lineBreakPositionsByContent[address]) {
+          self.cache.lineBreakPositionsByContent[address] = self.sourceMappingDecoder.getLinebreakPositions(self.editor.getFile(self.lastCompilationResult.data.sourceList[rawLocation.file]))  
+        }
+        var lineColumnPos = self.sourceMappingDecoder.convertOffsetToLineColumn(rawLocation, self.cache.lineBreakPositionsByContent[address])
+        self.highlight(lineColumnPos, rawLocation);
+      }
+    });
+  });
 }
 
-Debugger.prototype.debug = function (receipt) {
-  if (this.onDebugRequested) this.onDebugRequested();
+/**
+ * Start debugging using Remix
+ *
+ * @param {String} txHash    - hash of the transaction
+ */
+Debugger.prototype.debug = function (txHash) {
   var self = this;
-  this.debugger.web3().eth.getTransaction(receipt.transactionHash, function (error, tx) {
+  this.debugger.web3().eth.getTransaction(txHash, function (error, tx) {
     if (!error) {
       self.debugger.debug(tx);
     }
   });
 };
 
+/**
+ * highlight the given @arg lineColumnPos
+ *
+ * @param {Object} lineColumnPos - position of the source code to hightlight {start: {line, column}, end: {line, column}}
+ * @param {Object} rawLocation - raw position of the source code to hightlight {start, length, file, jump}
+ */
+Debugger.prototype.highlight = function (lineColumnPos, rawLocation) {
+  var name = utils.fileNameFromKey(this.editor.getCacheFile()); // current opened tab
```

Comment:
Could also be a different file, but I think it's better to do that in a separate PR because we first need the ability to switch tabs without triggering recompilation.